src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/services/TaskManagementService.java

```diff
@@ -222,44 +333,66 @@ public void cancelTask(String taskId,
+            accessControlResponse.getUserInfo().getUid()
+        );
 
-        Map<String, CamundaVariable> variables = camundaService.getTaskVariables(taskId);
+        if (isRelease2EndpointsFeatureEnabled) {
+            TaskResource taskResource = roleAssignmentVerificationWithAssigneeCheckAndHierarchy(
+                taskId,
+                userId,
+                accessControlResponse,
+                permissionsRequired);
+
+            taskHasCompleted = taskResource.getState().equals(CFTTaskState.COMPLETED);
+            //Safe-guard
+            if (taskResource.getAssignee() == null) {
+                throw new TaskStateIncorrectException(
+                    String.format("Could not complete task with id: %s as task was not previously assigned", taskId)
+                );
+            }
+        } else {
+            CamundaTask camundaTask = camundaService.getUnmappedCamundaTask(taskId);
 
-        List<PermissionTypes> permissionsRequired = asList(OWN, EXECUTE);
-        roleAssignmentVerificationWithAssigneeCheckAndHierarchy(
-            camundaTask.getAssignee(),
-            userId,
-            variables,
-            accessControlResponse.getRoleAssignments(),
-            permissionsRequired
-        );
+            //Safe-guard
+            if (camundaTask.getAssignee() == null) {
+                throw new TaskStateIncorrectException(
+                    String.format("Could not complete task with id: %s as task was not previously assigned", taskId)
+                );
+            }
 
-        boolean isFeatureEnabled = launchDarklyFeatureFlagProvider.getBooleanValue(
+            Map<String, CamundaVariable> variables = camundaService.getTaskVariables(taskId);
+            // Check that task state was not already completed
+            String taskState = camundaService.getVariableValue(variables.get(TASK_STATE.value()), String.class);
+            taskHasCompleted = TaskState.COMPLETED.value().equals(taskState);
+            roleAssignmentVerificationWithAssigneeCheckAndHierarchy(
+                camundaTask.getAssignee(),
+                userId,
+                variables,
+                accessControlResponse.getRoleAssignments(),
+                permissionsRequired
+            );
+        }
+        final boolean isFeatureEnabled = launchDarklyFeatureFlagProvider.getBooleanValue(
             FeatureFlag.RELEASE_2_CANCELLATION_COMPLETION_FEATURE,
             userId
         );
 
-        if (isFeatureEnabled) {
+        if (isFeatureEnabled || isRelease2EndpointsFeatureEnabled) {
             //Lock & update Task
             TaskResource task = findByIdAndObtainLock(taskId);
             task.setState(CFTTaskState.COMPLETED);
             //Perform Camunda updates
-            camundaService.completeTask(taskId, variables);
+            camundaService.completeTask(taskId, taskHasCompleted);
             //Commit transaction
             cftTaskDatabaseService.saveTask(task);
```

Comment:
If taskHasCompleted returns true, then we don't need to save it again. I believe its unnecessary db call.